ascon-aead/src/asconcore.rs

```diff
@@ -0,0 +1,379 @@
+}
+
+/// Parameters of an Ascon instance
+pub(crate) trait Parameters {
+    /// Size of the secret key
+    ///
+    /// For internal use-only.
+    type KeySize: ArrayLength<u8>;
+    /// Internal storage for secret keys
+    ///
+    /// For internal use-only.
+    type InternalKey: InternalKey<Self::KeySize>;
+
+    /// Number of bytes to process per round
+    const COUNT: usize;
+    /// Initialization vector used to initialize Ascon's state
+    ///
+    /// For internal use-only
+    const IV: u64;
+}
+
+/// Parameters for Ascon-128
+pub(crate) struct Parameters128;
+
+impl Parameters for Parameters128 {
+    type KeySize = U16;
+    type InternalKey = InternalKey16;
+
+    const COUNT: usize = 8;
+    const IV: u64 = 0x80400c0600000000;
+}
+
+/// Parameters for Ascon-128a
+pub(crate) struct Parameters128a;
+
+impl Parameters for Parameters128a {
+    type KeySize = U16;
+    type InternalKey = InternalKey16;
+
+    const COUNT: usize = 16;
+    const IV: u64 = 0x80800c0800000000;
+}
+
+/// Parameters for Ascon-80pq
+pub(crate) struct Parameters80pq;
+
+impl Parameters for Parameters80pq {
+    type KeySize = U20;
+    type InternalKey = InternalKey20;
+
+    const COUNT: usize = 8;
+    const IV: u64 = 0xa0400c0600000000;
+}
+
+/// Core implementation of Ascon for one encryption/decryption operation
+pub(crate) struct AEADCore<'a, P: Parameters> {
+    state: State,
+    key: &'a P::InternalKey,
+}
+
+impl<'a, P: Parameters> AEADCore<'a, P> {
+    pub(crate) fn new(internal_key: &'a P::InternalKey, nonce: &GenericArray<u8, U16>) -> Self {
+        let mut state = State::new(
+            if P::KeySize::USIZE == 20 {
+                P::IV ^ internal_key.get_k0()
+            } else {
+                P::IV
+            },
+            internal_key.get_k1(),
+            internal_key.get_k2(),
+            u64_from_be_bytes(&nonce[..8]),
+            u64_from_be_bytes(&nonce[8..]),
+        );
+
+        state.permute_12();
+        if P::KeySize::USIZE == 20 {
+            state[2] ^= internal_key.get_k0();
+        }
+        state[3] ^= internal_key.get_k1();
+        state[4] ^= internal_key.get_k2();
+
+        Self {
+            state,
+            key: internal_key,
+        }
+    }
+
+    /// Permutation with 12 rounds and application of the key at the end
+    fn permute_12_and_apply_key(&mut self) {
+        self.state.permute_12();
+        self.state[3] ^= self.key.get_k1();
+        self.state[4] ^= self.key.get_k2();
+    }
+
+    /// Permutation with 6 or 8 rounds based on the parameters
+    #[inline(always)]
+    fn permute_state(&mut self) {
+        if P::COUNT == 8 {
+            self.state.permute_6();
+        } else {
+            self.state.permute_8();
+        }
+    }
+
+    fn process_associated_data(&mut self, mut associated_data: &[u8]) {
+        if !associated_data.is_empty() {
+            // TODO: rewrite with as_chunks once stabilized
+            // https://github.com/rust-lang/rust/issues/74985
```

Comment:
Absent `as_chunks` it still seems like you could use `chunks_exact`?

Reply:
Yes. Thanks for the hint.